### PR TITLE
Make the <~> operator equivalent to `channel.read`/`channel.write`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR makes the `<~>` IO operator throw, and insofar as possible equivalent in semantics to `channel.read` and `channel.write`. It also adds the `throws` keyword to `dsi` IO routines for domains and arrays, as a consequence of the above.

If a SystemError is thrown from within IO routines up through `readThis`/`writeThis`, then the error is caught and the equivalent QIO error code is stored in the channel. Otherwise, the error is unpacked and the `EIO` error code is stored in the channel.

As of this time there is no way to distinguish between `SystemError` thrown by the user and `SystemError` thrown by the Chapel IO machinery.

The test `test/classes/deitz/class/tree.chpl` was rewritten to use an iterative postorder traversal instead of a recursive iterator due to compilation bugs (likely due to #7134).

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none` and `CHPL_LLVM=llvm`
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet` and `CHPL_LLVM=llvm`

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
FYI, I plan to figure out why the test `classes/deitz/classes/tree` is segfaulting sometime early this week.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -3168,35 +3182,49 @@ private inline proc _write_one_internal(_channel_internal:qio_channel_ptr_t,
...
+    etmp = qio_channel_error(_channel_internal);
+    if !etmp {
+      etmp = _write_one_internal(_channel_internal, kind, x, origLocale);
+      _qio_channel_set_error_unlocked(_channel_internal, etmp);
     }
+    error = etmp;
     this.unlock();
   }
+
+  if !error then
```
Comment (dlongnecke-cray) [inline modules/standard/IO.chpl]: I think this should be "if error"?

Comment (dlongnecke-cray) [reply modules/standard/IO.chpl]: Mistake on my part - forgot that `QIO_ERROR_NONE = 0`. Logically, this statement is saying "if there was an error, then...".

**Comment by dlongnecke-cray** [comment]
Outstanding questions before review:

- Do we want to store `Error` objects in the channel instead of QIO error codes? Right now we unpack SystemErrors and for all other subclasses of error store a general `IOError` in the channel (which isn't very descriptive).
- Do we want to rethrow errors in addition to stuffing error codes in the channel?
- Several if not all of the IO routines are marked `inline`. Ideally, that inline tag would be removed where possible.

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -3230,17 +3258,29 @@ inline proc channel.readwrite(ref x) where !this.writing {
...
+
+    //
+    // TODO: Ideally, we would like to replace calls to `channel.readIt`
+    // entirely with calls to `channel.read`, however ioLiterals seem to
+    // expect the channel error state to be set _immediately_ after a call
+    // to `_read_one_internal`, and `channel.read` was throwing a EFORMAT
+    // error when called instead.
+    // So for now, I've adjusted `channel.readIt` (and `channel.writeIt`)
+    // to be throwing.
+    //
```
Comment (mppf) [inline modules/standard/IO.chpl]: I don't think we should keep the above comment

File: test/classes/deitz/class/tree.chpl [hunk 3]
```diff
@@ -10,14 +10,52 @@ class BinaryTree {
   }
 }
 
-iter BinaryTree.postOrder(): eltType {
+//
+// Old code uses a recursive iterator, which has trouble compiling when
+// called by writeln now that the <~> operator throws. See private issue
+// #644.
```
Comment (mppf) [inline test/classes/deitz/class/tree.chpl]: It seems this should be a public issue if it's:
 * an issue a user could encounter
 * something we want to refer to in the test system

In any case, I think this is just https://github.com/chapel-lang/chapel/issues/7134

Comment (dlongnecke-cray) [reply test/classes/deitz/class/tree.chpl]: Thanks, I had no idea about #7134.

**Pull request merged**